# IMDB 영화 리뷰 감성분석 - LSTM

In [1]:
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from tensorflow.keras.datasets import imdb
import warnings
warnings.filterwarnings('ignore')

In [2]:
(X_train, y_train), (X_test, y_test) = imdb.load_data()
X_train.shape, X_test.shape, y_train.shape, y_test.shape

17473536/17464789 [==============================] - 0s 0us/step


((25000,), (25000,), (25000,), (25000,))

- 전부 다 25000개의 데이터셋을 가짐
- 이러면 x_test를 10000건은 valid, 15000건은 test셋으로 분리
- X_train = (25000,)

In [3]:
print('등장한 단어수', len(imdb.get_word_index()))
print('리뷰 최대 길이', max(len(s) for s in X_train))
print('리뷰 평균 길이', sum(map(len, X_train)) / len(X_train))

1654784/1641221 [==============================] - 0s 0us/step
등장한 단어수 88584
리뷰 최대 길이 2494
리뷰 평균 길이 238.71364


### LSTM으로 IMDB 리뷰 감성 분석
- 단어 빈도수 : 10,000개 ( 총 88,584)
- 문장의 단어수 : 500( 최대 2,494)
- test data 중 40%(10,000개)는 검증용으로 분리

In [4]:
import numpy as np
import tensorflow as tf
seed = 2022
np.random.seed(seed)
tf.random.set_seed(seed)

In [5]:
from tensorflow.keras.models import Sequential, load_model
from tensorflow.keras.layers import Embedding, LSTM, Dense
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.callbacks import ModelCheckpoint, EarlyStopping

In [6]:
(X_train, y_train), (X_test, y_test) = imdb.load_data(num_words = 10000)
X_train.shape, X_test.shape, y_train.shape

((25000,), (25000,), (25000,))

In [7]:
max_len = 500
X_train = pad_sequences(X_train, maxlen=max_len)
X_test = pad_sequences(X_test, maxlen=max_len)
X_train.shape, X_test.shape
# 500으로 맞춰져 있다.

((25000, 500), (25000, 500))

In [8]:
from sklearn.model_selection import train_test_split
X_test, X_valid, y_test, y_valid = train_test_split(
    X_test, y_test, stratify=y_test, test_size=0.4, random_state=seed
)
X_test.shape, X_valid.shape, y_test.shape, y_valid.shape

((15000, 500), (10000, 500), (15000,), (10000,))

- Case 1) LSTM
    - embedding dim : 100
    - hidden_units : 128
    

In [9]:
model1 = Sequential([ 
    Embedding(10000, 100, input_length=max_len),
    LSTM(128),
    Dense(1, activation='sigmoid')
])
model1.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, 500, 100)          1000000   
                                                                 
 lstm (LSTM)                 (None, 128)               117248    
                                                                 
 dense (Dense)               (None, 1)                 129       
                                                                 
Total params: 1,117,377
Trainable params: 1,117,377
Non-trainable params: 0
_________________________________________________________________


In [10]:
model1.compile('adam', 'binary_crossentropy', ['accuracy'])

model_path = 'best-imdb-lstm.h5'
mc = ModelCheckpoint(model_path, save_best_only=True, verbose=1)
es = EarlyStopping(patience=5)

In [11]:
hist1 = model1.fit(
    X_train, y_train, epochs=30, batch_size=64, 
    validation_data=[X_valid, y_valid],
    callbacks=[mc,es]
)

Epoch 1/30
391/391 [==============================] - ETA: 0s - loss: 0.5041 - accuracy: 0.7448
Epoch 00001: val_loss improved from inf to 0.32594, saving model to best-imdb-lstm.h5
391/391 [==============================] - 46s 102ms/step - loss: 0.5041 - accuracy: 0.7448 - val_loss: 0.3259 - val_accuracy: 0.8625
Epoch 2/30
391/391 [==============================] - ETA: 0s - loss: 0.2730 - accuracy: 0.8916
Epoch 00002: val_loss did not improve from 0.32594
391/391 [==============================] - 40s 101ms/step - loss: 0.2730 - accuracy: 0.8916 - val_loss: 0.3520 - val_accuracy: 0.8481
Epoch 3/30
391/391 [==============================] - ETA: 0s - loss: 0.1987 - accuracy: 0.9248
Epoch 00003: val_loss did not improve from 0.32594
391/391 [==============================] - 39s 101ms/step - loss: 0.1987 - accuracy: 0.9248 - val_loss: 0.3304 - val_accuracy: 0.8594
Epoch 4/30
391/391 [==============================] - ETA: 0s - loss: 0.2824 - accuracy: 0.8811
Epoch 00004: val_loss did 

In [12]:
best_model = load_model(model_path)
best_model.evaluate(X_test, y_test)

469/469 [==============================] - 18s 37ms/step - loss: 0.3199 - accuracy: 0.8683


[0.319904625415802, 0.8683333396911621]